### ISSUE TITLE
Seeking advice for using deepcopy-gen

### ISSUE THREAD

**Issue by hongchaodeng** [kubernetes/code-generator#4]
I tried to use deepcopy-gen on custom resource, but it doesn't work: 1. no xxx.deepcopy.go file was generated; 2. saw many confusing errors in log message.

Created this issue to seek help and advice for usng deepcopy-gen.

Here's what I did:

- Add tag to my custom resource:
	```Go
	// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

	// EtcdClusterList is a list of etcd clusters.
	type EtcdClusterList struct {
		metav1.TypeMeta `json:",inline"`
		// Standard list metadata
		// More info: http://releases.k8s.io/HEAD/docs/devel/api-conventions.md#metadata
		metav1.ListMeta `json:"metadata,omitempty"`
		Items           []EtcdCluster `json:"items"`
	}

	// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object

	type EtcdCluster struct {
		metav1.TypeMeta   `json:",inline"`
		metav1.ObjectMeta `json:"metadata,omitempty"`
		Spec              ClusterSpec   `json:"spec"`
		Status            ClusterStatus `json:"status"`
	}
	```
  For more information, check out https://github.com/coreos/etcd-operator/blob/master/pkg/apis/etcd/v1beta1/cluster.go .
- Compile deepcopy-gen from code-generator repo at a8e1019df0df00506fe1db5232a2af30ff4c87aa
- Run command:
  ```
  deepcopy-gen -i github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta1 --go-header-file="$GOPATH/src/k8s.io/kubernetes/hack/boilerplate/boilerplate.go.txt" -v=4 --logtostderr --bounding-dirs "github.com/coreos/etcd-operator/pkg/apis" --output-file-base zz_generated.deepcopy
  ```

No generated file comes out.
Here's the log message:
https://gist.github.com/hongchaodeng/2a8725a8c5b314aa2e401c72d119f0ce

Let me know if you need more details.

Some resources:
- https://github.com/openshift/origin/blob/2ff4203c2979623908b83bbb05f1f542bf567e56/cmd/service-catalog/go/src/github.com/kubernetes-incubator/service-catalog/Makefile#L175-L181

**Comment by tamalsaha** [comment]
Hi, I have learning about code generator recently. I think I know the issue. I saw that etcd operator uses client-go 4.0.0. This repo works with client-go master. You need to use code-generators from the Kubernetes 1.7 branch.

This is what I have done with client-go 4.0 and worked for me:

- Build a docker image from Kube 1.7 https://github.com/appscode/gengo-builder/blob/master/Dockerfile
- Run code-gen using docker image https://github.com/appscode/voyager/blob/master/hack/codegen.sh

**Comment by hongchaodeng** [comment]
@tamalsaha 
Thanks for your help.
We want to bump up dependencies to align with upcoming k8s 1.8 .
So I update dependencies in etcd-operator to latest client-go:

```
- package: k8s.io/client-go
  version: c6f8cf2c47d21d55fa0df928291b2580544886c8
- package: k8s.io/api
  version: 926789af7ddda62752e08bc9b93f1d1ebbc7b2de
- package: k8s.io/apimachinery
  version: a28c9b0b532638dcfb2c8b1b6f8f1e4a21041752
```

Still the same error.

**Comment by hongchaodeng** [comment]
@tamalsaha 
One question:
I'm confused how to ensure the dependency is deterministic here:
https://github.com/appscode/gengo-builder/blob/f1e9a54e21dc712442a61fb7674d46410ab3b01a/Dockerfile#L4

**Comment by tamalsaha** [comment]
My knowledge is limited regarding dependency. What I understood was that you need to use the correct version of the generator so that the **generated code uses the correct dependencies**. To compile the generator you only need the kubernetes project.

**Comment by hongchaodeng** [comment]
Still failed..

I tried to generate a docker image from the same Dockerfile: https://github.com/appscode/gengo-builder/blob/f1e9a54e21dc712442a61fb7674d46410ab3b01a/Dockerfile

Then run similar commands:
```

docker run --rm -ti \
  -v "${GOPATH}/src/github.com/coreos/etcd-operator":"${DOCKER_REPO_ROOT}" \
  -w "$DOCKER_REPO_ROOT" \
  "${IMAGE}" deepcopy-gen \
  --go-header-file="hack/docker/k8s_codegen/boilerplate.go.txt" \
  --input-dirs="github.com/coreos/etcd-operator/pkg/apis/etcd/v1beta1" \
  --v=4 --logtostderr \
  --bounding-dirs "github.com/coreos/etcd-operator/pkg/apis" \
  --output-file-base zz_generated.deepcopy
```

still fail:
```
I0919 04:17:51.887042       1 parse.go:417] type checker: /usr/local/go/src/crypto/tls/cipher_suites.go:18:2: could not import golang_org/x/crypto/chacha20poly1305 (unable to import "golang_org/x/crypto/chacha20poly1305": cannot find package "golang_org/x/crypto/chacha20poly1305" in any of:
	/go/src/github.com/coreos/etcd-operator/vendor/golang_org/x/crypto/chacha20poly1305 (vendor tree)
	/usr/local/go/src/golang_org/x/crypto/chacha20poly1305 (from $GOROOT)
	/go/src/golang_org/x/crypto/chacha20poly1305 (from $GOPATH))
I0919 04:17:51.900739       1 parse.go:417] type checker: /usr/local/go/src/crypto/tls/key_agreement.go:20:2: could not import golang_org/x/crypto/curve25519 (unable to import "golang_org/x/crypto/curve25519": cannot find package "golang_org/x/crypto/curve25519" in any of:
	/go/src/github.com/coreos/etcd-operator/vendor/golang_org/x/crypto/curve25519 (vendor tree)
	/usr/local/go/src/golang_org/x/crypto/curve25519 (from $GOROOT)
	/go/src/golang_org/x/crypto/curve25519 (from $GOPATH))
I0919 04:17:51.918757       1 parse.go:366] type checking encountered some issues in "crypto/tls", but ignoring.
I0919 04:17:51.968790       1 parse.go:417] type checker: /usr/local/go/src/net/http/h2_bundle.go:46:2: could not import golang_org/x/net/http2/hpack (unable to import "golang_org/x/net/http2/hpack": cannot find package "golang_org/x/net/http2/hpack" in any of:
	/go/src/github.com/coreos/etcd-operator/vendor/golang_org/x/net/http2/hpack (vendor tree)
	/usr/local/go/src/golang_org/x/net/http2/hpack (from $GOROOT)
	/go/src/golang_org/x/net/http2/hpack (from $GOPATH))
I0919 04:17:51.970830       1 parse.go:417] type checker: /usr/local/go/src/net/http/h2_bundle.go:47:2: could not import golang_org/x/net/idna (unable to import "golang_org/x/net/idna": cannot find package "golang_org/x/net/idna" in any of:
	/go/src/github.com/coreos/etcd-operator/vendor/golang_org/x/net/idna (vendor tree)
	/usr/local/go/src/golang_org/x/net/idna (from $GOROOT)
	/go/src/golang_org/x/net/idna (from $GOPATH))
I0919 04:17:51.976932       1 parse.go:417] type checker: /usr/local/go/src/net/http/h2_bundle.go:48:2: could not import golang_org/x/net/lex/httplex (unable to import "golang_org/x/net/lex/httplex": cannot find package "golang_org/x/net/lex/httplex" in any of:
	/go/src/github.com/coreos/etcd-operator/vendor/golang_org/x/net/lex/httplex (vendor tree)
	/usr/local/go/src/golang_org/x/net/lex/httplex (from $GOROOT)
	/go/src/golang_org/x/net/lex/httplex (from $GOPATH))
```

**Comment by hongchaodeng** [comment]
I got it worked..
I create a `doc.go` file under the same dir:
```
// +k8s:deepcopy-gen=package

package v1beta1
```
Same command.

Thanks @tamalsaha for your help all the same. Appreciated!

**Comment by lidajia** [comment]
I tried to use deepcopy-gen on custom resource with k8s 1.8.3, but get some errors. I have created the doc.go.  I'm seeking for some advice
here is the error info:
I1205 11:28:10.736185    6980 parse.go:417] type checker: E:\Go\src\runtime\trace.go:150:30: cannot use (traceBufHeader literal) (value of type traceBufHeader) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.757187    6980 parse.go:417] type checker: E:\Go\src\runtime\trace.go:150:7: array length 64 << 10 - unsafe.Sizeof((traceBufHeader literal)) (value of type uintptr) must be constant
I1205 11:28:10.757187    6980 parse.go:417] type checker: E:\Go\src\runtime\mgcwork.go:303:37: cannot use (workbufhdr literal) (value of type workbufhdr) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.757187    6980 parse.go:417] type checker: E:\Go\src\runtime\mgcwork.go:303:7: array length (_WorkbufSize - unsafe.Sizeof((workbufhdr literal))) / sys.PtrSize (value of type uintptr) must be constant
I1205 11:28:10.758187    6980 parse.go:417] type checker: E:\Go\src\runtime\chan.go:27:28: cannot use (hchan literal) (value of type hchan) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.758187    6980 parse.go:417] type checker: E:\Go\src\runtime\chan.go:27:66: cannot use (hchan literal) (value of type hchan) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.759189    6980 parse.go:417] type checker: E:\Go\src\runtime\chan.go:27:14: unsafe.Sizeof((hchan literal)) + uintptr(-int(unsafe.Sizeof((hchan literal))) & (maxAlign - 1)) (value of type uintptr) is not constant
I1205 11:28:10.761186    6980 parse.go:417] type checker: E:\Go\src\runtime\hashmap.go:80:31: cannot use (struct{b bmap; v int64} literal).v (value of type int64) as unsafe.ArbitraryType value in argument to unsafe.Offsetof
I1205 11:28:10.761186    6980 parse.go:417] type checker: E:\Go\src\runtime\hashmap.go:80:15: unsafe.Offsetof((struct{b bmap; v int64} literal).v) (value of type uintptr) is not constant
I1205 11:28:10.763186    6980 parse.go:417] type checker: E:\Go\src\runtime\mheap.go:148:47: cannot use (mcentral literal) (value of type mcentral) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.764189    6980 parse.go:417] type checker: E:\Go\src\runtime\mheap.go:148:13: array length sys.CacheLineSize - unsafe.Sizeof((mcentral literal)) % sys.CacheLineSize (value of type uintptr) must be constant
I1205 11:28:10.765187    6980 parse.go:417] type checker: E:\Go\src\runtime\mfinal.go:28:65: cannot use (finalizer literal) (value of type finalizer) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.765187    6980 parse.go:417] type checker: E:\Go\src\runtime\mfinal.go:28:11: array length (_FinBlockSize - 2 * sys.PtrSize - 2 * 4) / unsafe.Sizeof((finalizer literal)) (value of type uintptr) must be constant
I1205 11:28:10.767187    6980 parse.go:417] type checker: E:\Go\src\runtime\mheap.go:1535:41: cannot use (gcBitsHeader literal) (value of type gcBitsHeader) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.768188    6980 parse.go:417] type checker: E:\Go\src\runtime\mheap.go:1535:27: unsafe.Sizeof((gcBitsHeader literal)) (value of type uintptr) is not constant
I1205 11:28:10.768188    6980 parse.go:417] type checker: E:\Go\src\runtime\mstats.go:436:39: cannot use memstats.by_size (variable of type [67]struct{size uint32; nmalloc uint64; nfree uint64}) as unsafe.ArbitraryType value in argume
nt to unsafe.Offsetof
I1205 11:28:10.769186    6980 parse.go:417] type checker: E:\Go\src\runtime\mstats.go:436:76: cannot use memstats.by_size[0] (variable of type struct{size uint32; nmalloc uint64; nfree uint64}) as unsafe.ArbitraryType value in argumen
t to unsafe.Sizeof
I1205 11:28:10.769186    6980 parse.go:417] type checker: E:\Go\src\runtime\panic.go:120:34: cannot use (_defer literal) (value of type _defer) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.770190    6980 parse.go:417] type checker: E:\Go\src\runtime\panic.go:120:20: unsafe.Sizeof((_defer literal)) (value of type uintptr) is not constant
I1205 11:28:10.772187    6980 parse.go:417] type checker: E:\Go\src\runtime\sema.go:51:42: cannot use (semaRoot literal) (value of type semaRoot) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.773188    6980 parse.go:417] type checker: E:\Go\src\runtime\sema.go:51:8: array length sys.CacheLineSize - unsafe.Sizeof((semaRoot literal)) (value of type uintptr) must be constant
I1205 11:28:10.781187    6980 parse.go:366] type checking encountered some issues in "runtime", but ignoring.
I1205 11:28:10.782187    6980 parse.go:417] type checker: E:\Go\src\sync\map.go:69:31: cannot convert new(interface{}) (value of type *interface{}) to unsafe.Pointer
I1205 11:28:10.783187    6980 parse.go:417] type checker: E:\Go\src\sync\pool.go:68:27: cannot use (poolLocalInternal literal) (value of type poolLocalInternal) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:10.783187    6980 parse.go:417] type checker: E:\Go\src\sync\pool.go:68:7: array length 128 - unsafe.Sizeof((poolLocalInternal literal)) % 128 (value of type uintptr) must be constant
I1205 11:28:10.783187    6980 parse.go:366] type checking encountered some issues in "sync", but ignoring.
I1205 11:28:19.704699    6980 parse.go:417] type checker: E:\Go\src\crypto\cipher\xor.go:12:36: cannot use uintptr(0) (constant 0 of type uintptr) as unsafe.ArbitraryType value in argument to unsafe.Sizeof
I1205 11:28:19.705698    6980 parse.go:417] type checker: E:\Go\src\crypto\cipher\xor.go:12:18: int(unsafe.Sizeof(uintptr(0))) (value of type int) is not constant
I1205 11:28:19.706699    6980 parse.go:366] type checking encountered some issues in "crypto/cipher", but ignoring.
I1205 11:28:21.787401    6980 parse.go:417] type checker: E:\Go\src\crypto\tls\cipher_suites.go:18:2: could not import golang_org/x/crypto/chacha20poly1305 (unable to import "golang_org/x/crypto/chacha20poly1305": cannot find package
"golang_org/x/crypto/chacha20poly1305" in any of:
        E:\Go\src\golang_org\x\crypto\chacha20poly1305 (from $GOROOT)
        E:\workspace\src\golang_org\x\crypto\chacha20poly1305 (from $GOPATH))
I1205 11:28:21.840404    6980 parse.go:417] type checker: E:\Go\src\crypto\tls\key_agreement.go:20:2: could not import golang_org/x/crypto/curve25519 (unable to import "golang_org/x/crypto/curve25519": cannot find package "golang_org/
x/crypto/curve25519" in any of:
        E:\Go\src\golang_org\x\crypto\curve25519 (from $GOROOT)
        E:\workspace\src\golang_org\x\crypto\curve25519 (from $GOPATH))
I1205 11:28:21.845403    6980 parse.go:366] type checking encountered some issues in "crypto/tls", but ignoring.
I1205 11:28:22.522440    6980 parse.go:417] type checker: E:\Go\src\net\http\h2_bundle.go:46:2: could not import golang_org/x/net/http2/hpack (unable to import "golang_org/x/net/http2/hpack": cannot find package "golang_org/x/net/http
2/hpack" in any of:
        E:\Go\src\golang_org\x\net\http2\hpack (from $GOROOT)
        E:\workspace\src\golang_org\x\net\http2\hpack (from $GOPATH))
I1205 11:28:22.523439    6980 parse.go:417] type checker: E:\Go\src\net\http\h2_bundle.go:47:2: could not import golang_org/x/net/idna (unable to import "golang_org/x/net/idna": cannot find package "golang_org/x/net/idna" in any of:
        E:\Go\src\golang_org\x\net\idna (from $GOROOT)
        E:\workspace\src\golang_org\x\net\idna (from $GOPATH))
I1205 11:28:22.523439    6980 parse.go:417] type checker: E:\Go\src\net\http\h2_bundle.go:48:2: could not import golang_org/x/net/lex/httplex (unable to import "golang_org/x/net/lex/httplex": cannot find package "golang_org/x/net/lex/
httplex" in any of:
        E:\Go\src\golang_org\x\net\lex\httplex (from $GOROOT)
        E:\workspace\src\golang_org\x\net\lex\httplex (from $GOPATH))
I1205 11:28:22.575442    6980 parse.go:417] type checker: E:\Go\src\net\http\transport.go:32:2: could not import golang_org/x/net/proxy (unable to import "golang_org/x/net/proxy": cannot find package "golang_org/x/net/proxy" in any of
:
        E:\Go\src\golang_org\x\net\proxy (from $GOROOT)
        E:\workspace\src\golang_org\x\net\proxy (from $GOPATH))
I1205 11:28:22.593443    6980 parse.go:366] type checking encountered some issues in "net/http", but ignoring.
I1205 11:28:28.073469    6980 main.go:81] Completed successfully.
